File: src/test/java/net/sf/rails/game/state/ManagerTest.java
package net.sf.rails.game.state;

import static org.junit.Assert.*;

import net.sf.rails.game.state.Item;
import net.sf.rails.game.state.Manager;
import net.sf.rails.game.state.Root;

import org.junit.Before;
import org.junit.Test;

public class ManagerTest {

    private static final String MANAGER_ID = "manager";
    private static final String ITEM_ID = "item";
    private static final String ANOTHER_ITEM_ID = "anotherItem";
    private static final String ANOTHER_MANAGER_ID = "anotherManager";

    private Root root;
    private Manager manager, anotherManager;
    private Item item, anotherItem;

    @Before
    public void setUp() {
        root = Root.create();
        item = new AbstractItemImpl(root, ITEM_ID);
        manager = new ManagerImpl(item, MANAGER_ID);
        anotherItem = new AbstractItemImpl(manager, ANOTHER_ITEM_ID);
        anotherManager = new ManagerImpl(manager, ANOTHER_MANAGER_ID);
    }

    @Test
    public void testGetId() {
        assertEquals(MANAGER_ID, manager.getId());
        assertEquals(ANOTHER_MANAGER_ID, anotherManager.getId());
    }

    @Test
    public void testGetParent() {
        assertSame(item, manager.getParent());
        assertSame(manager, anotherManager.getParent());
    }

    @Test
    public void testGetContext() {
        assertSame(root, manager.getContext());
        assertSame(manager, anotherManager.getContext());
    }

    @Test
    public void testGetURI() {
        assertEquals(ITEM_ID + Item.SEP + MANAGER_ID, manager.getURI());
        assertEquals(ANOTHER_MANAGER_ID, anotherManager.getURI());
    }

    @Test
    public void testGetFullURI() {
        assertEquals(Item.SEP + ITEM_ID + Item.SEP + MANAGER_ID, manager.getFullURI());
        assertEquals(Item.SEP + ITEM_ID + Item.SEP + MANAGER_ID + Item.SEP + ANOTHER_MANAGER_ID, anotherManager.getFullURI());
    }

    @Test
    public void testLocate() {
        // anotherItem is local
        assertSame(anotherItem, manager.locate(anotherItem.getURI()));
        assertSame(anotherItem, manager.locate(anotherItem.getFullURI()));
        
        // item is not local
        assertNull(manager.locate(item.getURI()));
        assertSame(item, manager.locate(item.getFullURI()));

        // manager is not local in itself, but in root
        assertNull(manager.locate(manager.getURI()));
        assertSame(manager, root.locate(manager.getURI()));
        assertSame(manager, manager.locate(manager.getFullURI()));

        // anotherManager is not local in itself, but in manager
        assertNull(anotherManager.locate(anotherManager.getURI()));
        assertSame(anotherManager, manager.locate(anotherManager.getURI()));
        assertSame(anotherManager, anotherManager.locate(anotherManager.getFullURI()));
    }

    @Test
    public void testGetRoot() {
        assertSame(root, manager.getRoot());
        assertSame(root, anotherManager.getRoot());
    }

}

File: dotnet/GameStateTest/ManagerTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class ManagerTest
    {
        private const string MANAGER_ID = "manager";
        private const string ITEM_ID = "item";
        private const string ANOTHER_ITEM_ID = "anotherItem";
        private const string ANOTHER_MANAGER_ID = "anotherManager";

        private Root root;
        private Manager manager, anotherManager;
        private IItem item, anotherItem;

        [TestInitialize]
        public void SetUp()
        {
            root = Root.Create();
            item = new AbstractItemImpl(root, ITEM_ID);
            manager = new ManagerImpl(item, MANAGER_ID);
            anotherItem = new AbstractItemImpl(manager, ANOTHER_ITEM_ID);
            anotherManager = new ManagerImpl(manager, ANOTHER_MANAGER_ID);
        }

        [TestMethod]
        public void TestGetId()
        {
            Assert.AreEqual(MANAGER_ID, manager.Id);
            Assert.AreEqual(ANOTHER_MANAGER_ID, anotherManager.Id);
        }

        [TestMethod]
        public void TestGetParent()
        {
            Assert.AreSame(item, manager.Parent);
            Assert.AreSame(manager, anotherManager.Parent);
        }

        [TestMethod]
        public void TestGetContext()
        {
            Assert.AreSame(root, manager.GetContext);
            Assert.AreSame(manager, anotherManager.GetContext);
        }

        [TestMethod]
        public void TestGetURI()
        {
            Assert.AreEqual(ITEM_ID + IItemConsts.SEP + MANAGER_ID, manager.URI);
            Assert.AreEqual(ANOTHER_MANAGER_ID, anotherManager.URI);
        }

        [TestMethod]
        public void TestGetFullURI()
        {
            Assert.AreEqual(IItemConsts.SEP + ITEM_ID + IItemConsts.SEP + MANAGER_ID, manager.FullURI);
            Assert.AreEqual(IItemConsts.SEP + ITEM_ID + IItemConsts.SEP + MANAGER_ID + IItemConsts.SEP + ANOTHER_MANAGER_ID, anotherManager.FullURI);
        }

        [TestMethod]
        public void TestLocate()
        {
            // anotherItem is local
            Assert.AreSame(anotherItem, manager.Locate(anotherItem.URI));
            Assert.AreSame(anotherItem, manager.Locate(anotherItem.FullURI));

            // item is not local
            Assert.IsNull(manager.Locate(item.URI));
            Assert.AreSame(item, manager.Locate(item.FullURI));

            // manager is not local in itself, but in root
            Assert.IsNull(manager.Locate(manager.URI));
            Assert.AreSame(manager, root.Locate(manager.URI));
            Assert.AreSame(manager, manager.Locate(manager.FullURI));

            // anotherManager is not local in itself, but in manager
            Assert.IsNull(anotherManager.Locate(anotherManager.URI));
            Assert.AreSame(anotherManager, manager.Locate(anotherManager.URI));
            Assert.AreSame(anotherManager, anotherManager.Locate(anotherManager.FullURI));
        }

        [TestMethod]
        public void testGetRoot()
        {
            Assert.AreSame(root, manager.GetRoot);
            Assert.AreSame(root, anotherManager.GetRoot);
        }

    }
}
